sql/core/src/main/scala/org/apache/spark/sql/connector/read/V1Scan.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read
+
+import org.apache.spark.annotation.{Experimental, Unstable}
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.sql.sources.BaseRelation
+
+/**
+ * A trait that should be implemented by V1 DataSources that would like to leverage the DataSource
+ * V2 read code paths.
+ *
+ * This interface is designed to provide Spark DataSources time to migrate to DataSource V2 and
+ * will be removed in a future Spark release.
+ *
+ * @since 3.0.0
+ */
+@Experimental
+@Unstable
+trait V1Scan extends Scan {
+
+  /**
+   * Creates an `BaseRelation` that can scan data from DataSource v1 to RDD[Row]. The returned
+   * relation must be a `TableScan` instance.
```

Comment:
Why doesn't this return `TableScan` if the requirement is that one is returned?

Also, the source connector API is written in Java. Why is this in Scala?

Reply:
`TableScan` is a mixin trait, what we need is `BaseRelation with TableScan`, but that doesn't work well for java.

Reply:
I write it in Scala to follow `V1WriteBuilder`. It's another topic if we should write these fallback traits in Java or Scala, I follow the existing choice to be consistent.

Reply:
Maybe we missed the other class. The connector API is in Java, so I think this should be as well.

In Java, you can use a method parameter, like this:
```java
<T extends BaseRelation & TableScan> T toV1TableScan();
```